File: HW2 POS Tagging/viterbi.py
import sys
import re
import math
import itertools
from pprint import pprint
from collections import defaultdict

INIT_STATE = 'init'
FINAL_STATE = 'final'
OOV_SYMBOL = 'OOV'

hmmfile = 'my.hmm'
inputfile = 'ptb.22.txt'

tags = set()  # i.e. K in the slides, a set of unique POS tags
trans = {}  # transisions
emit = {}  # emissions
voc = {}  # encountered words

"""
This part parses the my.hmm file you have generated and obtain the transition and emission values.
"""
with open(hmmfile) as hmmfile:
  for line in hmmfile.read().splitlines():
    # extract information for transition and emission
    trans_reg = 'trans\s+(\S+)\s+(\S+)\s+(\S+)'
    emit_reg = 'emit\s+(\S+)\s+(\S+)\s+(\S+)'
    trans_match = re.match(trans_reg, line)
    emit_match = re.match(emit_reg, line)
    if trans_match:
      qq, q, p = trans_match.groups()
      trans[(qq, q)] = math.log(float(p))
      tags.update([qq, q])
    elif emit_match:
      q, w, p = emit_match.groups()
      emit[(q, w)] = math.log(float(p))
      voc[w] = 1
      tags.update([q])

"""
This part parses the file with the test sentences, and runs the sentence through the viterbi algorithm.
"""
with open(inputfile) as inputfile:
  for line in inputfile.read().splitlines():
    line = line.split(' ')
    # initialize pi.
    # i.e. set pi(0, *, *) = 1 from slides
    pi = {(0, INIT_STATE): 0.0}  # 0.0 because using logs
    bp = {}  # backpointers

    # for each word in sentence and their index
    for k, word in enumerate(line):
      k = k + 1
      if word not in voc:
        # change unseen words into OOV, since OOV is assigned a score in train_hmm. This will give these unseen words a score instead of a mismatch.
        word = OOV_SYMBOL
      for u, v in itertools.product(tags, tags):  # python nested for loop
        # i.e. the first bullet point from the slides.
        # Calculate the scores (p) for each possible combinations of (u, v)
        if (v, u) in trans and (u, word) in emit and (k - 1, v) in pi:
          p = pi[(k - 1, v)] + trans[(v, u)] + emit[(u, word)]
          if (k, u) not in pi or p > pi[(k, u)]:
            # here, fine the max of all the calculated p, update it in the pi dictionary
            pi[(k, u)] = p
            # also keeping track of the backpointer
            bp[(k, u)] = v

    # second bullet point from the slides.
    # Taking the case for the last word. Find the corrsponding POS tag for that word so we can then start the backtracing.
    foundgoal = False
    goal = float('-inf')
    tag = INIT_STATE
    for v in tags:
      # You want to try each (tag, FINAL_STATE) pair for the last word and find which one has max p. That will be the tag you choose.
      if (v, FINAL_STATE) in trans and (len(line), v) in pi:
        p = pi[(len(line), v)] + trans[(v, FINAL_STATE)]
        if not foundgoal or p > goal:
          # finding tag with max p
          goal = p
          foundgoal = True
          tag = v

    # if pi is greater than previous recorded pi
    if foundgoal:
      # y is the sequence of final chosen tags
      y = []
      for i in xrange(len(line), 1, -1):  # counting from the last word
        # bp[(i, tag)] gives you the tag for word[i - 1].
        # we use that and traces through the tags in the sentence.
        y.append(bp[(i, tag)])
        tag = bp[(i, tag)]

      # y is appened last tag first. Reverse it.
      y.reverse()
      # print the final output
      print ' '.join(y)
    else:
      # append blank line if something fails so that each sentence is still printed on the correct line.
      print '\n'

File: HW2 POS Tagging/viterbi.pl
#!/usr/bin/perl

# Noah A. Smith
# 2/21/08
# Runs the Viterbi algorithm (no tricks other than logmath!), given an
# HMM, on sentences, and outputs the best state path.

# Usage:  viterbi.pl hmm-file < text > tags

# The hmm-file should include two kinds of lines.  One is a transition:
# trans Q R P
# where Q and R are whitespace-free state names ("from" and "to," 
# respectively) and P is a probability.  The other kind of line is an 
# emission:
# emit Q S P
# where Q is a whitespace-free state name, S is a whitespace-free
# emission symbol, and P is a probability.  It's up to you to make sure
# your HMM properly mentions the start state (named init by default),
# the final state (named final by default) and out-of-vocabulary
# symbols (named OOV by default).

# If the HMM fails to recognize a sequence, a blank line will be written.
# Change $verbose to 1 for more verbose output.

# special keywords:
#  $init_state   (an HMM state) is the single, silent start state
#  $final_state  (an HMM state) is the single, silent stop state
#  $OOV_symbol   (an HMM symbol) is the out-of-vocabulary word

use bytes;

$init_state = "init";
$final_state = "final";
$OOV_symbol = "OOV";

$verbose = 0;

# read in the HMM and store the probabilities as log probabilities

$hmmfile = shift;
open(HMM, "<$hmmfile") or die "could not open $hmmfile";
while(<HMM>) {
    if(($qq, $q, $p) = (m/trans\s+(\S+)\s+(\S+)\s+(\S+)/)) {
	$A{$qq}{$q} = log($p);
	$States{$qq} = 1;
	$States{$q} = 1;
    }
    elsif(($q, $w, $p) = (m/emit\s+(\S+)\s+(\S+)\s+(\S+)/)) {
	$B{$q}{$w} = log($p);
	$States{$q} = 1;
	$Voc{$w} = 1;
    }
}
close(HMM);

while(<>) { # read in one sentence at a time
    chomp;
    @w = split;
    $n = scalar(@w);
    unshift @w, "";
    %V = ();
    %Backtrace = ();
    $V{0}{$init_state} = 0.0;  # base case of the recurisve equations!
    for($i = 1; $i <= $n; ++$i) {     # work left to right ...
	# if a word isn't in the vocabulary, rename it with the OOV symbol
	unless(defined $Voc{$w[$i]}) {
	    print STDERR "OOV:  $w[$i]\n" if($verbose);
	    $w[$i] = $OOV_symbol;
	}
	foreach $q (keys %States) { # consider each possible current state
	    foreach $qq (keys %States) { # each possible previous state
		if(defined $A{$qq}{$q}  # only consider "non-zeros"
		   and defined $B{$q}{$w[$i]} 
		   and defined $V{$i - 1}{$qq}) 
		{
		    $v = $V{$i - 1}{$qq} + $A{$qq}{$q} + $B{$q}{$w[$i]};
		    if(!(defined $V{$i}{$q}) or $v > $V{$i}{$q}) {
			# if we found a better previous state, take note!
			$V{$i}{$q} = $v;  # Viterbi probability
			$Backtrace{$i}{$q} = $qq; # best previous state
		    }
		}
	    }
	    print STDERR "V[$i, $q] = $V{$i}{$q} ($B{$i}{$q})\n" if($verbose);
	}
    }
    # this handles the last of the Viterbi equations, the one that brings
    # in the final state.
    $foundgoal = 0;
    foreach $qq (keys %States) { # for each possible state for the last word
	if(defined $A{$qq}{$final_state} and defined $V{$n}{$qq}) {
	    $v = $V{$n}{$qq} + $A{$qq}{$final_state};
	    if(!$foundgoal or $v > $goal) {
		# we found a better path; remember it
		$goal = $v;
		$foundgoal = 1;
		$q = $qq;
	    }
	}
    }
    
    # this is the backtracking step.
    if($foundgoal) {
	@t = ();
	for($i = $n; $i > 0; --$i) {
	    unshift @t, $q;
	    $q = $Backtrace{$i}{$q};
	}
    }
    print STDERR exp($goal), "\n" if($verbose);
    if($foundgoal) { print join " ", @t; }
    print "\n";
}
